File: tests/examples/trdf_main.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void c_calobjf(int n, double *x, double *f, int *flag) {

  // Problem HS37 from Hock-Schittkowski collection

  *flag = 0;

  *f = - x[0] * x[1] * x[2];

  return;

}

void c_calcon(int n, double *x, int ind, double *c, int *flag) {

  // Problem HS37 from Hock-Schittkowski collection

  *flag = 0;

  if ( ind == 0 ) {

    *c = - 72.0 + x[0] + 2.0 * x[1] + 2.0 * x[2];

  }
  else if ( ind == 1 ) {

    *c = - x[0] - 2.0 * x[1] - 2.0 * x[2];

  }
  else {

    *flag = -1;

  }

  return;

}

void c_caljac(int n, double *x, int ind, int *jcvar, double *jcval,
	    int *jcnnz, int lim, _Bool *lmem, int *flag) {

  // Problem HS37 from Hock-Schittkowski collection

  *flag = 0;

  *lmem = 0;

  if ( ind == 0 ) {

    *jcnnz = 3;
    jcvar[0] = 0;
    jcval[0] = 1.0;
    jcvar[1] = 1;
    jcval[1] = 2.0;
    jcvar[2] = 2;
    jcval[2] = 2.0;

  } else if ( ind == 1 ) {

    *jcnnz = 3;
    jcvar[0] = 0;
    jcval[0] = - 1.0;
    jcvar[1] = 1;
    jcval[1] = - 2.0;
    jcvar[2] = 2;
    jcval[2] = - 2.0;

  } else {
    
    *flag = -1;

  }

  return;

}

void c_calhc(int n, double *x, int ind, int *hcrow, int *hccol, double *hcval,
	   int *hcnnz, int lim, _Bool *lmem, int *flag) {
  

  // Problem HS37 from Hock-Schittkowski collection

   *flag = 0;

   *lmem = 0;

   *hcnnz = 0;

   return;

}


int main()
{

  int i, fcnt, m, n;

  double f, feas, *x, *l, *u;

  _Bool *equatn, *linear, ccoded[2];

  // Problem HS37 from Hock-Schittkowski collection

  // Number of variables
  n = 3;

  // Allocates initial point and bounds

  x = (double*) malloc(n * sizeof(double));
  l = (double*) malloc(n * sizeof(double));
  u = (double*) malloc(n * sizeof(double));
  
  if ( x == NULL || l == NULL || u == NULL ) {

    printf("\nMemory allocation error.\n");

    return 1;

  }

  for (i = 0; i < n; i++) {

    l[i] = 0.0;
    u[i] = 42.0;
    x[i] = 10.0;

  }

  // Number of constraints
  m = 2;

  equatn = (_Bool*) malloc(m * sizeof(_Bool));
  linear = (_Bool*) malloc(m * sizeof(_Bool));

  if ( equatn == NULL || linear == NULL ) {

    printf("\nMemory allocation error.\n");

    return 1;

  }

  // Type of constraints

  for (i = 0; i < m; i++) equatn[i] = 0;
  for (i = 0; i < m; i++) linear[i] = 1;

  // Coded subroutines for constraints' derivatives

  ccoded[0] = 1;
  ccoded[1] = 1;


  // Calls the solver
  easytrdf(n,x,l,u,m,equatn,linear,ccoded,&c_calobjf,&c_calcon,
           &c_caljac,&c_calhc,&f,&feas,&fcnt);

  // Frees memory
  free(x);
  free(l);
  free(u);
  free(equatn);
  free(linear);

}

File: tests/examples/trdf_main.f
      PROGRAM PRINCIPAL

#include "tr_params.par"

C     LOCAL ARRAYS
      LOGICAL ccoded(2),equatn(MMAX),linear(MMAX)
      double precision X(NMAX),XL(NMAX),XU(NMAX)

C     USER-DEFINED SUBROUTINES
      external calobjf,calcon,caljac,calhc

C     NUMBER OF VARIABLES
      N= 3

C     NUMBER OF CONSTRAINTS
      M = 2

C     TYPE OF CONSTRAINTS
      do i = 1,M
         equatn(i) = .false.
      end do
      
      do i = i,m
         linear(i) = .true.
      end do

C     INITIAL POINT AND BOX CONSTRAINTS.
      DO 6 I=1,3
         X(I)=10.D0
         XL(I)= 0D0
 6       XU(I)= 42D0 

C     CODED SUBROUTINES FOR CONSTRAINTS' DERIVATIVES
      ccoded(1) = .true.
      ccoded(2) = .true.

C     CALLS THE ALGORITHM

      CALL EASYTRDF(N,X,XL,XU,M,EQUATN,LINEAR,CCODED,CALOBJF,CALCON,
     +              CALJAC,CALHC,F,FEAS,FCNT)

      END PROGRAM PRINCIPAL

C     ******************************************************************
C     ******************************************************************

      subroutine calobjf(n,x,f,flag)

!     SCALAR ARGUMENTS
      integer flag,n
      double precision f

!     ARRAY ARGUMENTS
      double precision x(n)

!     Problem HS37 from Hock-Schittkowski collection

      flag = 0

      f = - x(1) * x(2) * x(3)

      end subroutine calobjf

C     ******************************************************************
C     ******************************************************************

      subroutine calcon(n,x,ind,c,flag)

!     SCALAR ARGUMENTS
      integer ind,flag,n
      double precision c

!     ARRAY ARGUMENTS
      double precision x(n)

!     Problem HS37 from Hock-Schittkowski collection

      flag = 0

      if ( ind .eq. 1 ) then
         c = - 72.D0 + x(1) + 2.D0 * x(2) + 2.D0 * x(3)
         return
      elseif ( ind .eq. 2 ) then
         c = - x(1) - 2.D0 * x(2) - 2.D0 * x(3)
         return
      end if

      flag = -1

      end subroutine calcon

C     ******************************************************************
C     ******************************************************************

      subroutine caljac(n,x,ind,jcvar,jcval,jcnnz,lim,lmem,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,ind,jcnnz,lim,n

C     ARRAY ARGUMENTS
      integer jcvar(lim)
      double precision x(n),jcval(lim)

      lmem = .false.

      flag = 0

!     Problem HS37 from Hock-Schittkowski collection

      if ( ind .eq. 1 ) then
         jcnnz    = 3
         jcvar(1) = 1
         jcval(1) = 1.D0  
         jcvar(2) = 2
         jcval(2) = 2.D0  
         jcvar(3) = 3
         jcval(3) = 2.D0  
         return
      else if ( ind .eq. 2 ) then
         jcnnz    = 3
         jcvar(1) = 1
         jcval(1) = -1.D0
         jcvar(2) = 2
         jcval(2) = -2.D0     
         jcvar(3) = 3
         jcval(3) = -2.D0     
         return
      else
         
         flag = -1

      end if
  
      end subroutine caljac

C     ******************************************************************
C     ******************************************************************

      subroutine calhc(n,x,ind,hcrow,hccol,hcval,hcnnz,lim,lmem,flag)

      implicit none

C     SCALAR ARGUMENTS
      logical lmem
      integer flag,hcnnz,ind,lim,n

C     ARRAY ARGUMENTS
      integer hccol(lim),hcrow(lim)
      double precision hcval(lim),x(n)

!     Problem HS37 from Hock-Schittkowski collection

      flag = 0

      lmem = .false.

!     The problem has linear constraints

      hcnnz = 0

      end
